cli/internal/run/run.go

```diff
@@ -263,17 +252,62 @@ func (c *RunCommand) runOperation(g *completeGraph, rs *runSpec, backend *api.La
 		}
 	}
 
-	if rs.Opts.stream {
-		c.Ui.Output(fmt.Sprintf("%s %s %s", ui.Dim("• Running"), ui.Dim(ui.Bold(strings.Join(rs.Targets, ", "))), ui.Dim(fmt.Sprintf("in %v packages", rs.FilteredPkgs.Len()))))
+	engine, err := buildTaskGraph(&g.TopologicalGraph, g.Pipeline, rs)
+	if err != nil {
+		c.Ui.Error(fmt.Sprintf("Error preparing engine: %s", err))
+		return 1
 	}
-	// TODO(gsoltis): I think this should be passed in, and close called from the calling function
-	// however, need to handle the graph case, which early-returns
-	runState := NewRunState(rs.Opts, startAt)
-	runState.Listen(c.Ui, time.Now())
-	engine := core.NewScheduler(&g.TopologicalGraph)
-	colorCache := NewColorCache()
-	var logReplayWaitGroup sync.WaitGroup
-	for taskName, value := range g.Pipeline {
+	exitCode := 0
+	if rs.Opts.dotGraph != "" {
+		err := c.generateDotGraph(engine.TaskGraph, filepath.Join(rs.Opts.cwd, rs.Opts.dotGraph))
+		if err != nil {
+			c.logError(c.Config.Logger, "", err)
+			return 1
+		}
+	} else if rs.Opts.dryRun {
+		tasksRun, err := c.executeDryRun(engine)
+		if err != nil {
+			c.logError(c.Config.Logger, "", err)
+			return 1
+		}
+		packagesInScope := rs.FilteredPkgs.UnsafeListOfStrings()
+		sort.Strings(packagesInScope)
+		if rs.Opts.dryRunJson {
+			dryRun := make(map[string][]string)
+			dryRun["packages"] = packagesInScope
+			dryRun["tasks"] = tasksRun
+			bytes, err := json.MarshalIndent(dryRun, "", "  ")
+			if err != nil {
+				c.logError(c.Config.Logger, "", errors.Wrap(err, "failed to render to JSON"))
+				return 1
+			}
+			c.Ui.Output(string(bytes))
+		} else {
+			c.Ui.Info("Packages in scope:")
+			for _, pkg := range packagesInScope {
+				c.Ui.Output(fmt.Sprintf(ui.Bold("• %v"), pkg))
+			}
+			c.Ui.Info("Tasks to run:")
+			for _, task := range tasksRun {
```

Comment:
Do we know hashes here? can we print them out?

Reply:
We can calculate them and print them out. Include them in JSON too? It would probably be a list of objects:
```json5
[
  {
    "task": "foo#build",
    "hash": "abcdabcd"
  },
  ...
]
```

Reply:
sgtm